### PR TITLE
Add back SSL_use_certificate_chain_file

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Addresses `CryptoAlg-2165`

### Description of changes: 
Librelp consumes this so we have to add it back. I translated the underlying function to C++ while I was at it. The function ([like it's OpenSSL version)](https://github.com/openssl/openssl/blob/fe487609c17dac049f867f230e09ee090b65e966/ssl/ssl_rsa.c#L437) has some `SSL_CTX` and `SSL` specific logic, so there needs to be a way to differentiate. 

### Call-outs:
N/A

### Testing:
New test for `SSL_use_certificate_chain_file` and `SSL_CTX_use_certificate_chain_file`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: ssl/ssl_test.cc [hunk 1]
```diff
@@ -11810,5 +11816,46 @@ TEST_P(PerformHybridHandshakeTest, PerformHybridHandshake) {
   }
 }
 
+// This test opens and writes a temporary pem file to test certificate file
+// loading. It is disabled by default since we may not always have permissions
+// to create files on some platforms or the generated files may encounter race
+// conditions when running multiple ssl_tests on different threads. This is
+// reenabled via all_tests.go where suitable.
```
Comment (WillChilds-Klein) [inline ssl/ssl_test.cc:11823]: we have [other tests using `tmpfile()`](https://github.com/aws/aws-lc/commit/7285f2ec75bd8d8426474dccbc1674b921fe576a) that write files and haven't (to my knowledge) encountered permissions issues or race conditions. randomly generated file names will solve your concurrency issues; if `tmpfile()` doesn't resolve your permissions issues, you might be able to `mmap` your way around the issue.

File: ssl/ssl_file.cc [hunk 2]
```diff
@@ -475,32 +475,40 @@ int SSL_CTX_use_PrivateKey_file(SSL_CTX *ctx, const char *file, int type) {
...
+  if (in == nullptr) {
     OPENSSL_PUT_ERROR(SSL, ERR_R_BUF_LIB);
-    goto end;
   }
 
-  if (BIO_read_filename(in, file) <= 0) {
+  if (BIO_read_filename(in.get(), file) <= 0) {
     OPENSSL_PUT_ERROR(SSL, ERR_R_SYS_LIB);
-    goto end;
   }
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply ssl/ssl_file.cc:498]: Ah my bad I forgot to add the `return 0` when I translated this to C++.

File: ssl/ssl_file.cc [hunk 3]
```diff
@@ -510,23 +518,31 @@ int SSL_CTX_use_certificate_chain_file(SSL_CTX *ctx, const char *file) {
...
 
-    while ((ca = PEM_read_bio_X509(in, NULL, ctx->default_passwd_callback,
-                                   ctx->default_passwd_callback_userdata)) !=
-           NULL) {
-      r = SSL_CTX_add0_chain_cert(ctx, ca);
-      if (!r) {
+      if (temp_ret == 0) {
         X509_free(ca);
         ret = 0;
-        goto end;
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply ssl/ssl_file.cc:525]: yes, we do :)

File: ssl/ssl_test.cc (hunk 1 above)
Comment (samuel40791765) [reply ssl/ssl_test.cc:11823]: Ah thanks, I had no idea this existed! The Android issue that was mentioned in the test was the permissions issue I was running into, so this should help get past that.

File: ssl/CMakeLists.txt [hunk 4]
```diff
@@ -80,7 +80,7 @@ if(BUILD_TESTING)
     span_test.cc
     ssl_test.cc
     ssl_c_test.c
- )
+  )
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply ssl/CMakeLists.txt:83]: This was intentional, the original line spacing was off.

**Review by WillChilds-Klein — APPROVED** [review]

**Review by skmcgrail — APPROVED** [review]

**Review by justsmth — APPROVED** [review]

**Pull request merged**